README
NAME
    Apache::DBI::Cache - a DBI connection cache

SYNOPSIS
      use Apache::DBI::Cache debug=>3, bdb_env=>'/tmp/tmp',
                             plugin=>'Apache::DBI::Cache::mysql',
                             plugin=>['DBM', sub {...}, sub {...}],
                             ...;

INSTALLATION
    This installs 2 modules "Apache::DBI::Cache" und
    "Apache::DBI::Cache::mysql".

     perl Makefile.PL
     make
     make test
     make install

  Testing
   MySQL
    Since Apache::DBI::Cache::mysql depends on MySQL you need 2 databases on
    the same DB server listening on port 3306 to test it. Nothing will be
    written to them. Only connections are established and the "SHOW
    PROCESSLIST" command is executed.

    The connection parameters are passed to the test as environment
    variables. If you do not want to test it leave them empty and the plugin
    test ("t/100mysql") is silently skipped.

    These environment variables are used:

    MYSQL1, MYSQL2
        the 2 database names

    MYSQL_HOST
        the database host. If omitted "localhost" is used.

    MYSQL_USER, MYSQL_PASSWD
        you credentials.

    My "make test" command looks like:

     MYSQL1=dbitest1 MYSQL2=dbitest2 make test

   BerkeleyDB
    If "BerkeleyDB" is not available a few tests are skipped.

DEPENDENCIES
    BerkeleyDB is used if installed to make DBI handle statistics visible
    for the whole Apache process group instead of a single process.

    DBI 1.37

    perl 5.8.0

DESCRIPTION
    This module is an alternative to the famous Apache::DBI module. As
    Apache::DBI it provides persistent DBI connections.

    It can be used with mod_perl1, mod_perl2 and even standalone.

  WHY ANOTHER MODULE FOR THE SAME?
    Apache::DBI has a number of limitations. Firstly, it is not possible to
    get multiple connections with the same parameters. A common scenario for
    example is to use one connection to perform transactions and another to
    perform simple lookups in the same database. With Apache::DBI it is very
    likely to get the same connection if you mean to use different.

    With Apache::DBI all connections are reset at end of a request.

    Apache::DBI does not regard database specific functions to cache handles
    more aggressively. For example a mysql DSN can look like

     dbi:mysql:test:localhost:3306

    or

     dbi:mysql:host=localhost;db=test

    Both point to the same database but for Apache::DBI they are different.
    Apache::DBI::Cache recognizes these two by means of a *mysql* plugin.

    The plugin even recognizes connections to different databases on the
    same mysql server as the same connection and issues a "USE database"
    command before returning the actual handle to the user. Hence, with
    Apache::DBI::Cache many the overall number of connections to a DB server
    can be dramatically reduced.

  HOW DOES IT WORK?
    To decide whether to use Apache::DBI::Cache or not it is essential to
    know how it works. As with Apache::DBI Apache::DBI::Cache uses a hook
    provided by the DBI module to intercept "DBI->connect()" calls. Also do
    Apache::DBI::Cache maintain a cache of active handles.

    When a new connection is requested and the cache is empty a new
    connection is established and returned to the user. At this point it is
    not cached at all. Here Apache::DBI::Cache differs from Apache::DBI.

    Later either "disconnect" is called on the handle or it simply goes out
    of scope and the garbage collector calls a "DESTROY" method if provided.
    Both events are intercepted by Apache::DBI::Cache. Only then the handle
    is put in the cache.

    This means a handle is never really disconnected. "$dbh->{Active}" will
    always return true no matter how often "disconnect" is called. Further,
    you can prevent a handle from getting reused by simply not forgetting
    it.

  USAGE
    Different to Apache::DBI Apache::DBI::Cache must be "use"ed not
    "require"ed. That means it's "import" function must be called.

    When used with mod_perl (versions 1.x or 2.x) this is best done in a
    "startup.pl" or in a "<Perl>" section in the "httpd.conf". See mod_perl
    documentation for more information.

    Thereafter, "DBI->connect" is called as usual. No special treatment is
    needed.

    When Apache::Status or Apache2::Status is used Apache::DBI::Cache
    provides an extra menu item to show statistics on handles. The loading
    order of the Apache::Status and Apache::DBI::Cache is irrelevant.

  FUNCTIONS
    import - "use" parameter
        Parameters to the "use" statement are given in a "key => value"
        fashion.

         use Apache::DBI::Cache debug=>3, logger=>sub {...},
                                plugin=>['driver', sub {}, sub {}],
                                plugin=>'Apache::DBI::Cache::mysql',
                                use_bdb=>1,
                                bdb_env=>'/tmp/mybdbenv',
                                bdb_memcache=>20*1024,
                                ...;

        * plugin
          loads a plugin, see also "PLUGINS" below. The plugin can be
          specified as a 3-element array or by name. In the second case the
          "import" simply "use"s the module. This option can be given
          multiple times.

        * use_bdb, bdb_env and bdb_memcache
          Apache::DBI::Cache can use BerleleyDB as a shared memory
          implementation to maintain statistics for a group of processes
          instead of a single process.

          "use_bdb" specify whether to use BerkeleyDB or not. If ommitted
          Apache::DBI::Cache will try to load and use BerkeleyDB. If that
          fails it silently provides per process statistics. If "use_bdb" is
          true Apache::DBI::Cache dies if it cannot use BerkeleyDB. If
          "use_bdb" is false per process statistics are maintained and
          BerkeleyDB is not used.

          "bdb_env" specifies a path to a directory where BerkeleyDB can put
          it's temporary files. If omitted "/tmp/Apache::DBI::Cache" is
          used. The parent directory of this directory must exists and be
          writeable.

          "bdb_memcache" specifies the size of the shared memory segment
          that is allocated by BerkeleyDB. Depending on the number of
          handles in your configuration a few kilobytes are enough. If
          omitted 20 kB are used.

          "bdb_env" and "bdb_memcache" can also be specified by the
          "APACHE_DBI_CACHE_ENVPATH" and "APACHE_DBI_CACHE_CACHESIZE"
          environment variables.

        * debug
          set a debug level. Under mod_perl this is almost irrelevant, see
          "logger" below.

        * logger
          here a logger function can be specified. It is called with the
          message verbosity level as the first parameter. The remaining
          parameters are concatenated to build the actual message.

          Currently there are 2 verbosity levels used 1 and 2. 0 is reserved
          for real errors. 1 mentions that the module has been initialized.
          2 rattle off normal processing messages.

          Apache::DBI::Cache provides 2 logger functions. One is controlled
          by the "debug" level setting (see above). A message is printed to
          STDERR if it's level is equal or greater the current debug level.

          The other logger is used when running under mod_perl. It is mainly
          controlled by the Apache "LogLevel" setting. Messages at level 0
          are printed as "$log->error", level 1 as "$log->info" and level 2
          as "$log->debug". For level 2 messages additionally the current
          debug level is checked to be greater or equal 2.

        * delimiter
          Here the internal key delimiter can be changed. It defaults to
          "\1". Changing it is necessary only when your DSN, username or
          password contain it or to provide more readable debugging
          messages.

    statistics
        returns a reference to the statistics hash. If BerkeleyDB is used it
        is tied to BerkeleyDB::Btree.

    statistics_as_html
        returns a reference to an array of HTML fragments. If mod_perl and
        Apache::Status or Apache::Status2 is used the output of this
        function is shown under http://HOST/STATUS/URI?DBI_conn.

    plugin( 'name', \&mangle, \&setup )
        installs a new plugin, see "PLUGINS" below. If a plugin for the
        specified database type was already installed it is returned as a
        2-element list:

         ($old_mangle, $old_setup)=
           plugin( 'name', \&new_mangle, \&new_setup );

        If called with an name only the current plugin is returned:

         ($old_mangle, $old_setup)=plugin( 'name' );

        To delete a plugin call

         ($old_mangle, $old_setup)=plugin( 'name', undef, undef );

    connect_on_init
        call this function multiple times with parameters you would pass to
        "DBI->connect" before calling "Apache::DBI::Cache::init", i.e. in
        your "startup.pl". Then "init" will establish all these connections.

    init
        This function is called once per child process to initialize
        Apache::DBI::Cache. If mod_perl is used this is done automatically
        in a PerlChildInitHandler

    finish
        This function must be called before a process is going to terminate.
        Under mod_perl it is automatically called in a PerlChildExitHandler.

    undef_at_request_cleanup( \$dbh1, \$dbh2, ... )
        When an application uses global variables to store handles they
        probably won't be reused because a global variable is ..., well
        global. This can be fixed by explicitly undefining these handles at
        request cleanup or by using this function. It simply collects all
        handle references passed in between 2 calls to
        "Apache::DBI::Cache::request_cleanup". When
        "Apache::DBI::Cache::request_cleanup" is called all these handles
        are undefined. The first call to this function during a request
        cycle installs "Apache::DBI::Cache::request_cleanup" as
        PerlCleanupHandler.

        With mod_perl2 this requires the PerlOption "GlobalRequest" to be
        set:

         PerlOption +GlobalRequest

        in your httpd.conf.

    request_cleanup
        This is the PerlCleanupHandler. If Apache::DBI::Cache is used
        standalone the application can call it from time to time.

  EXPORT
    Nothing.

PLUGINS
    Plugins are used to modify the caching for certain database types. They
    can change the caching key, issue database commands just before a handle
    is returned to the user or prevent handle caching entirely for a
    database type.

    There can only be one plugin per database type at a time.

    A plugin registers itself by calling "Apache::DBI::Cache::plugin"
    passing 3 parameters. The first parameter is simply the name of the
    database type. It matches the DBI driver name. Thus, a MySQL plugin
    passes the string "mysql" since the corresponding DBI driver is named
    "DBD::mysql". Whereas a PostgreSQL plugin passes either "Pg" or "PgPP"
    depending on the driver actually used.

    The 2nd and 3rd parameters are CODE references that are called just
    before a connection is chosen from the cache or newly established and
    after the connection is made just before it is returned to the caller.
    The first function can mangle the connection parameters the second
    perform additional setup steps. Further, I will call them *mangle* and
    *setup*.

    Thus, a plugin is registered this way:

     Apache::DBI::Cache::plugin('Name', \&mangle, \&setup);

    Normally, it is implemented as a separate module according to the
    following template, see Apache::DBI::Cache::mysql for example:

     package Apache::DBI::Cache::DRIVER;

     use strict;

     BEGIN {
       die "Please load Apache::DBI::Cache before"
         unless defined &Apache::DBI::Cache::plugin;

       ...

       Apache::DBI::Cache::plugin
           (
            'DRIVER',
            sub {},
            sub {}
           );
     }

     1;

  Calling Conventions
    * mangle
       ($dsn, $user, $passwd, $attr, $ctx, $nocache)=
          mangle($dsn, $user, $passwd, $attr);

      *mangle* is called with almost the same parameters as the original
      call to "DBI->connect". The "dbi:DRIVER:" prefix is stripped from the
      DSN. It is expected to return similar values plus an arbitrary context
      that is later passed to *setup* and an optional "nocache" flag.

      If $nocache is true or *mangle* returns an empty list a new connection
      is made and the handle is directly passed to the caller without
      further processing. Also *setup* will not be called. Such a handle
      will not be cached on "disconnect" or "DESTROY".

      *mangle* can change all parameters. The MySQL plugin for example
      deletes the actual database name from the DSN, reformats it according
      to a standard format and adds the standard port if it is omitted. The
      database is put in the context.

    * setup
       $rc=setup($dbh, $dsn, $user, $passwd, $attr, $ctx);

      The *setup* function performs additional setup steps on $dbh. The
      MySQL plugin for example issues a "USE database" command using the
      database from the context.

      A connection is considered dead if *setup* returns false.

TODO
    * periodically ping all cached handles
    * correct statistics when a process is finishes without calling
    "finish()"
    * redirect BerkeleyDB errors to logger

SEE ALSO
    Apache::DBI::Cache::mysql
    Apache::DBI

AUTHOR
    Torsten Foertsch, <torsten.foertsch@gmx.net>

    With suggestions from

        Andreas Nolte < andreas dot nolte at bertelsmann dot de >

        Dietmar Hanisch < dietmar dot hanisch at bertelsmann dot de > and

        Ewald Hinrichs < ewald dot hinrichs at bertelsmann dot de >

SPONSORING
    Sincere thanks to Arvato Direct Services (http://www.arvato.com/) for
    sponsoring this module and providing a test platform with several
    thousand DBI connections.

COPYRIGHT AND LICENSE
    Copyright (C) 2005 by Torsten Foertsch

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl extension Apache::DBI::Cache.

0.06  Wed Nov 30 10:18:54 CET 2005
	- first public version




Makefile.PL
use 5.008;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'Apache::DBI::Cache',
    VERSION_FROM      => 'lib/Apache/DBI/Cache.pm', # finds $VERSION
    PREREQ_PM         => {
			  DBI=>'1.37',
			 }, # e.g., Module::Name => 1.1
    dist              => {
                          PREOP => './mk_README.sh',
                         },
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/Apache/DBI/Cache.pod', # retrieve abstract from module
       AUTHOR         => 'Torsten Foertsch <torsten.foertsch@gmx.net>') : ()),
);

sub MY::postamble {
  return <<'EOF';
README: lib/Apache/DBI/Cache.pm mk_README.sh
	./mk_README.sh

rpm: dist
	rpmbuild -bb perl-Apache-DBI-Cache.spec && \
	rpmbuild -bs perl-Apache-DBI-Cache.spec
EOF
}


lib/Apache/DBI/Cache/mysql.pm
package Apache::DBI::Cache::mysql;

use 5.008;
use strict;
use warnings;
no warnings 'uninitialized';

our $VERSION = '0.05';

BEGIN {
  die "Please load Apache::DBI::Cache before"
    unless defined &Apache::DBI::Cache::plugin;

  require DBD::mysql;

  Apache::DBI::Cache::plugin
      (
       'mysql',
       sub {			# Idx generator $ctx
	 my ($dsn, $user, $passwd, $attr)=@_;
	 $Apache::DBI::Cache::LOG->(3, "mysql plugin: got dsn=$dsn");
	 $attr={%{$attr||{}},
		'user' => $user,
		'password' => $passwd,
	       };
	 DBD::mysql->_OdbcParse($dsn, $attr,
				['database', 'host', 'port']);
	 if( exists $attr->{host} ) {
	   $attr->{port}=3306 unless( exists $attr->{port} );
	 }
	 $attr->{AutoCommit}=1 unless( exists $attr->{AutoCommit} );
	 $dsn=join(';',
		   map( {length $attr->{$_}
			 ? $_."=".$attr->{$_}
			 : ()} (qw/host port/)));
	 my $ctx={};
	 $ctx->{database}=$attr->{database} if( length $attr->{database} );
	 ($user, $passwd)=delete @{$attr}{qw/user password host port database/};

	 $Apache::DBI::Cache::LOG->(3, "mysql plugin: returning dsn=$dsn");
	 return ($dsn, $user, $passwd, $attr, $ctx);
       },
       sub {			# connection reinit: issue 'use db'
	 my ($dbh, $dsn, $user, $passwd, $attr, $ctx)=@_;
	 if( exists $ctx->{database} ) {
	   $Apache::DBI::Cache::LOG->(3, "mysql plugin: use database $ctx->{database}");
	   my $rc=eval {
	     $dbh->{mysql_auto_reconnect}=0;
	     $dbh->do('USE '.$ctx->{database});
	   };
	   $Apache::DBI::Cache::LOG->(0, "mysql: USE $ctx->{database} failed".($@?": $@":"\n"))
	     unless($rc);
	   return $rc;
	 } else {
	   $Apache::DBI::Cache::LOG->(3, "mysql plugin: no database specified");
	   return 1;
	 }
       }
      );
}

1;

__END__

=head1 NAME

Apache::DBI::Cache::mysql - a Apache::DBI::Cache plugin

=head1 SYNOPSIS

  use Apache::DBI::Cache plugin=>'Apache::DBI::Cache::mysql',
                         ...;
 or

  use Apache::DBI::Cache;
  use Apache::DBI::Cache::mysql;

=head1 DESCRIPTION

B<NOTE:> Read L<Apache::DBI::Cache> before.

C<DBD::mysql> allows many different DSN syntaxes for connecting to the
same database server and the same database. This plugin transforms them
to a standard format thus allowing better DBI handle caching. Further,
if C<port> is omitted the standard port 3306 is inserted. The actual database
is deleted from the DSN and replaced by a C<USE database> command.

So, DBI connects to the following DSNs as the same user
at different times are actually performed with the same DSN
C<host=server;port=3306>. Subsequently C<USE db[1-3]> commands are
issued before passing the handle to the caller.

  dbi:mysql:dbname=db1;host=server
  dbi:mysql:db2:server:3306
  dbi:mysql:port=3306;database=db3;host=server

Apache::DBI::Cache can cache them all under the same key. So only one
real database connection is needed instead of 3 without the plugin.

Furthermore, C<mysql_auto_reconnect> is turned off.

=head1 SEE ALSO

=over 4

=item L<Apache::DBI::Cache>

=back

=head1 AUTHOR

Torsten Foertsch, E<lt>torsten.foertsch@gmx.netE<gt>

With suggestions from

=over 4

=item Z<>

Andreas Nolte E<lt> andreas dot nolte at bertelsmann dot de E<gt>

=item Z<>

Dietmar Hanisch E<lt> dietmar dot hanisch at bertelsmann dot de E<gt> and

=item Z<>

Ewald Hinrichs E<lt> ewald dot hinrichs at bertelsmann dot de E<gt>

=back

=head1 SPONSORING

Sincere thanks to Arvato Direct Services (http://www.arvato.com/) for
sponsoring this module and providing a test platform with several
thousand DBI connections.

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2005 by Torsten Foertsch

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


=cut


lib/Apache/DBI/Cache.pm
package Apache::DBI::Cache;

use 5.008;
use strict;
use warnings;
no warnings 'uninitialized';

our $VERSION = '0.06';

BEGIN { eval { require Apache; } }
BEGIN { eval { require mod_perl2; require Apache2::Module; } }

our $DEBUG = 0;
our $LOG=sub {
  my $level=shift;

  my @l=localtime;
  my $prefix=sprintf( '%5d %d%02d%02d %02d%02d%02d '.__PACKAGE__."   ",
		      $$, $l[5]+1900, $l[4]+1, @l[3,2,1,0] );
  print STDERR $prefix, @_, "\n" if( $DEBUG >= $level );
};

if( exists $ENV{MOD_PERL_API_VERSION} and $ENV{MOD_PERL_API_VERSION}==2 ) {
  require Apache2::ServerUtil;
  require Apache2::Log;
  my $s = Apache2::ServerUtil->server;
  $s->push_handlers(PerlChildInitHandler => \&init);
  $s->push_handlers(PerlChildExitHandler => \&finish);
  $s->push_handlers(PerlChildInitHandler =>
		    sub {
		      Apache2::Status->menu_item('DBI_conn' => 'DBI connections',
						 \&statistics_as_html)
			  if Apache2::Module::loaded('Apache2::Status');
		    } );
  $LOG=sub {
    my $level=shift;

    my $log=$s->log;

    if( $level==0 ) {
      $log->error(__PACKAGE__.': '.join('', @_));
    } elsif( $level==1 ) {
      $log->info(__PACKAGE__.': '.join('', @_));
    } elsif( $DEBUG>=2 ) {
      $log->debug("$$: ".join('', @_));
    }
  };
} elsif( exists $INC{'Apache.pm'} ) {
  require Apache::Log;
  if( Apache->can('push_handlers') ) {
    Apache->push_handlers(PerlChildInitHandler => \&init);
    Apache->push_handlers(PerlChildExitHandler => \&finish);
    Apache->push_handlers(PerlChildInitHandler =>
			  sub {
			    Apache::Status->menu_item('DBI_conn' => 'DBI connections',
						      \&statistics_as_html)
				if( Apache->can('module') and              # really?
				    Apache->module('Apache::Status') );    # Apache::Status too?
			  } );
  }
  if( Apache->can('server') ) {
    $LOG=sub {
      my $level=shift;

      my $log=Apache->server->log;

      if( $level==0 ) {
	$log->error(__PACKAGE__.': '.join('', @_));
      } elsif( $level==1 ) {
	$log->info(__PACKAGE__.': '.join('', @_));
      } elsif( $DEBUG>=2 ) {
	$log->debug("$$: ".join('', @_));
      }
    };
  }
}

use DBI ();

require_version DBI 1.37;

our %Connected;    # cache for database handles
our @ChildConnect; # connections to be established when a new httpd child is created
our %STAT;	   # gather statistics. This will be tied to BerkeleyDB.
our $STATdb;	   # object %STAT is tied to
our %localSTAT;
our %plugin;

our $DELIMITER="\1";
our $GLOBAL_DESTROY;
our $PRIVATE='private_'.__PACKAGE__;

my $use_bdb;
my $envpath;
my $bdb_memcache;

sub import {
  my $class=shift;
  my %o;
  while( my ($k, $v)=splice @_, 0, 2 ) {
    $o{$k}=[] unless( exists $o{$k} );
    push @{$o{$k}}, $v;
  }

  $DEBUG=$o{debug}->[0] if(exists $o{debug});
  $LOG=$o{logger}->[0] if(exists $o{logger});
  $DELIMITER=$o{delimiter}->[0] if(exists $o{delimiter});

  if(exists $o{use_bdb}) {
    $use_bdb=$o{use_bdb}->[0];
  } elsif( eval {require BerkeleyDB; require File::Path;} ) {
    $use_bdb=1;
  }

  if( $use_bdb ) {
    require BerkeleyDB; require File::Path;

    if(exists $o{bdb_env}) {
      $envpath=$o{bdb_env}->[0];
    } elsif(exists $ENV{APACHE_DBI_CACHE_ENVPATH}) {
      $envpath=$ENV{APACHE_DBI_CACHE_ENVPATH};
    }
    $envpath='/tmp/'.__PACKAGE__ unless(length $envpath);
    File::Path::rmtree( $envpath );
    die "ERROR: Cannot remove $envpath: $!\n" if( -e $envpath );

    if(exists $o{bdb_memcache}) {
      $bdb_memcache=$o{bdb_memcache}->[0];
    } elsif(exists $ENV{APACHE_DBI_CACHE_CACHESIZE}) {
      $bdb_memcache=$ENV{APACHE_DBI_CACHE_CACHESIZE};
    }
    $bdb_memcache=20*1024 if( $bdb_memcache==0 );
  }

  if(exists $o{plugin}) {
    foreach my $v (@{$o{plugin}}) {
      if(ref $v eq 'ARRAY') {
	plugin(@{$v});
      } else {
	eval "use $v";
	die "$@" if $@;
      }
    }
  }
}

sub plugin {
  my $driver=shift;
  my $old=$plugin{$driver};
  if( @_==2 and ref $_[0] eq 'CODE' and ref $_[1] eq 'CODE' ) {
    $plugin{$driver}=[@_];
  } elsif( @_==2 and !defined $_[0] and !defined $_[1] ) {
    delete $plugin{$driver};
  }
  return @{$old||[]};
}

sub _statop {
  my $statIdx=shift;

  if( $STATdb ) {
    my $lock=$STATdb->cds_lock;
    my $stat=$STAT{$statIdx} || [0,0,0,0,0];
    my $lstat=$localSTAT{$statIdx} || [0,0,0,0,0];
    while( my ($i, $x)=splice @_, 0, 2 ) {
      $stat->[$i]+=$x;
      $lstat->[$i]+=$x;
    }
    $STAT{$statIdx}=$stat;
    $localSTAT{$statIdx}=$lstat;
    $lock->cds_unlock;
  } else {
    my $stat=$STAT{$statIdx} || [0,0,0,0,0];
    while( my ($i, $x)=splice @_, 0, 2 ) {
      $stat->[$i]+=$x;
    }
    $STAT{$statIdx}=$stat;
  }
}

{
  my $init=0;
  sub init {
    if( $init ) {
      $LOG->(2, "init: already initialized");
      return 1;
    }

    undef $GLOBAL_DESTROY;
    if( $use_bdb ) {
      $LOG->(1, "init: initializing BerkeleyDB environment at $envpath");
      unless( -d $envpath ) {
	File::Path::mkpath( $envpath );
	die "ERROR: Cannot create $envpath: $!\n" unless( -d $envpath );
      }

      my $env=BerkeleyDB::Env->new
	( -Home=>$envpath,
	  -Cachesize=>$bdb_memcache,
	  -ErrFile=>\*STDERR,
	  -ErrPrefix=>__PACKAGE__.' BerkeleyDB',
	  -Flags=>(&BerkeleyDB::DB_CREATE|
		   &BerkeleyDB::DB_INIT_CDB|
		   &BerkeleyDB::DB_INIT_MPOOL),
	);
      die "ERROR: Cannot create BerkeleyDB environment ($envpath): $BerkeleyDB::Error\n"
	unless( $env );

      $STATdb=tie( %STAT, 'BerkeleyDB::Btree',
		   -Filename=>'handles.db',
		   -Env=>$env,
		   -Flags=>&BerkeleyDB::DB_CREATE,
		);
      $STATdb->filter_store_value( sub {$_=join ':', @$_} );
      $STATdb->filter_fetch_value( sub {$_=[split ':', $_]} );
    } else {
      $LOG->(1, "init: working without BerkeleyDB");
    }

    # redirect connects to us
    $DBI::connect_via=__PACKAGE__.'::connect';
    # redirect &DBI::connect_cached to DBI::connect
    undef &DBI::connect_cached;
    *DBI::connect_cached=\&DBI::connect;
    my @l;
    for my $aref (@ChildConnect) {
      shift @$aref if( UNIVERSAL::isa( $aref->[0], __PACKAGE__ ) );
      push @l, DBI->connect(@$aref);
    }
    @ChildConnect=();
    @l=();
    $init=1;
    1;
  }

  sub finish {
    $GLOBAL_DESTROY=1;
    if( $STATdb ) {
      foreach (keys %localSTAT) {
	_statop( $_,
		 0, -$localSTAT{$_}->[0],   # decr. handle count
		 1, -$localSTAT{$_}->[1] ); # decr. free count
      }
    }
    %Connected=();

    if( $use_bdb ) {
      $LOG->(2, "finish: shutting down BerkeleyDB environment");
      undef $STATdb;
      untie %STAT;
    }

    $init=0;
    1;
  }
}

{
  my @undef_at_cleanup;
  sub undef_at_request_cleanup {
    my @l=grep {ref eq 'REF' or ref eq 'SCALAR'} @_;
    return unless( @l );
    $LOG->(2, "undef_at_request_cleanup: @{[map {${$_}} @l]}");
    unless( @undef_at_cleanup ) {
      if( exists $ENV{MOD_PERL_API_VERSION} and $ENV{MOD_PERL_API_VERSION}==2 ) {
	require Apache2::RequestUtil;
	Apache2::RequestUtil->request
	    ->push_handlers(PerlCleanupHandler=>\&request_cleanup);
      } elsif( exists $INC{'Apache.pm'} and Apache->can( 'push_handlers' ) ) {
	Apache->push_handlers(PerlCleanupHandler=>\&request_cleanup);
      }
    }
    push @undef_at_cleanup, @l;
  }
  sub request_cleanup {
    $LOG->(2, "request_cleanup:");
    foreach my $v (@undef_at_cleanup) {
      $LOG->(2, "    undefining ${$v}");
      undef ${$v};
    }
    @undef_at_cleanup=();
  }
}

sub connect_on_init {
  # provide a handler which creates all connections during server startup

  # store connections
  push @ChildConnect, [@_];
}


# the connect method called from DBI::connect

sub connect {
  my $class = shift;
  unshift @_, $class if ref $class;
  my $drh    = shift;
  my @args   = map { defined $_ ? $_ : "" } @_;

  unless( 3 == $#args and ref $args[3] eq "HASH" ) {
    @args=(@args[0..2], {});
  }

  my ($Idx, $statIdx, $ctx);
  if( exists $plugin{$drh->{Name}} ) {
    my @l=$plugin{$drh->{Name}}->[0]->(@args);
    if( @l ) {
      my $nocache;
      ($ctx, $nocache)=splice @l, 4, 2;
      @args=@l;
      return $drh->connect(@args) if( $nocache );
    } else {
      return $drh->connect(@args);
    }
  }
  my $dsn="dbi:$drh->{Name}:$args[0]";

  $Idx    =join $DELIMITER, $drh->{Name}, $args[0], $args[1], $args[2];
  $statIdx=join $DELIMITER, $drh->{Name}, $args[0], $args[1];

  # the hash-reference differs between calls even in the same
  # process, so de-reference the hash-reference 
  # should we default to '__undef__' or something for undef values?
  map { $Idx .= "$DELIMITER$_=" .
	  (defined $args[3]->{$_}
	   ? $args[3]->{$_}
	   : '');
      } sort keys %{$args[3]};

  if( exists $Connected{$Idx} ) {
    while( my $dbh=shift @{$Connected{$Idx}} ) {
      local $GLOBAL_DESTROY=2;
      if( eval{$dbh->ping} ) {
	if( exists $plugin{$drh->{Name}} ) {
	  unless( $plugin{$drh->{Name}}->[1]->($dbh, @args, $ctx) ) {
	    _statop( $statIdx,
		     4, 1,	# plugin failure
		     1, -1,	# decr. free count
		     0, -1 );	# decr. handle count
	    $LOG->(2, "reusing connection to '$Idx' failed due to plugin error");
	    undef $dbh;
	    next;
	  }
	}
	_statop( $statIdx,
		 2, 1,		# incr. usage count
		 1, -1 );	# decr. free count
	$LOG->(2, "reusing connection to '$Idx'");
	$dbh->{$PRIVATE}->{disconnected}=0;
	return $dbh;
      } else {
	_statop( $statIdx,
		 3, 1,		# ping failure
		 1, -1,		# decr. free count
		 0, -1 );	# decr. handle count
	$LOG->(2, "reusing connection to '$Idx' failed due to PING failure");
	undef $dbh;
      }
    }
  }

  my $dbh=$drh->connect(@args);

  if( defined $dbh ) {
    $dbh->{$PRIVATE}=+{
		       disconnected=>0,
		       idx=>$Idx,
		       statIdx=>$statIdx,
		       attr=>$args[3],
		      };
    bless $dbh=>__PACKAGE__.'::db';

    if( exists $plugin{$drh->{Name}} ) {
      local $GLOBAL_DESTROY=2;
      unless( $plugin{$drh->{Name}}->[1]->($dbh, @args, $ctx) ) {
	_statop( $statIdx, 4, 1 ); # plugin error
	$LOG->(2, "new connection to '$Idx' failed due to plugin error");
	undef $dbh;
	return;
      }
    }

    _statop( $statIdx, 0, 1, 2, 1 ); # incr. handle count, incr. usage
  }

  # return the new database handle
  $LOG->(2, "new connection to '$Idx'");
  return $dbh;
}


sub statistics {
  return \%STAT;
}

{
  my %esc=(qw/" &quot; < &lt; > &gt; & &amp;/);
  $esc{' '}='&nbsp;';
  my $esc=sub {
    my $v=shift;
    $v=length $v ? $v : ' ';
    $v=~s/(["<>& ])/$esc{$1}/ge;
    $v;
  };

  sub statistics_as_html {
    my @s;

    my $lock;

    if( $STATdb ) {
      $lock=$STATdb->cds_lock;
      push @s, "<h1>DBI Handle Statistics for this machine</h1>\n";
    } else {
      push @s, "<h1>DBI Handle Statistics for process ".$esc->($$)."</h1>\n";
    }

    push @s, ('<table border="1">'."\n",
	      '<tr><th>Driver</th><th>Datasource</th>'.
	      '<th>Username</th><th>Handle Count</th><th>Free Handles</th>'.
	      '<th>Usage Count</th><th>Ping Failures</th><th>Plugin Failures</th>'.
	      '</tr>'."\n");
    foreach my $k (keys %STAT) {
      my $v=join( '</td><td>', map {$esc->($_)} @{$STAT{$k}} );
      push( @s,
	    ('<tr><td>'.
	     join('</td><td>',
		  map {$esc->($_)} (split /\Q$DELIMITER\E/, $k)[0..2]).
	     "</td><td>$v</td></tr>\n") );
    }
    push @s, "</table>\n";

    $lock->cds_unlock if( defined $lock );
    return \@s;
  }
}


# this is almost copied from Apache::DBI
# patch from Tim Bunce: Apache::DBI will not return a DBD ref cursor
@Apache::DBI::Cache::st::ISA = ('DBI::st');

# overload disconnect
{
  package Apache::DBI::Cache::db;
  use base qw(DBI::db);
  use strict;

  sub disconnect {
    my $dbh=shift;

    my $priv=$dbh->{$PRIVATE};
    my $Idx=$priv->{idx};

    $LOG->(2, "disconnect $Idx");

    if( $priv->{'disconnected'} ) {
      $LOG->(2, "already disconnect");
      return 1;
    }

    if( $dbh->{Active} and !$dbh->{AutoCommit} and eval {$dbh->rollback} ) {
      $LOG->(2, "ROLLBACK");
    }

    foreach my $k (keys %{$priv->{attr}}) {
      $dbh->{$k}=$priv->{attr}->{$k};
    }

    $priv->{'disconnected'}=1;

    Apache::DBI::Cache::_statop( $priv->{statIdx}, 1, 1 ); # incr. free count
    if( exists $Connected{$Idx} ) {
      push @{$Connected{$Idx}}, $dbh;
    } else {
      $Connected{$Idx}=[$dbh];
    }

    1;
  }

  sub DESTROY {
    my $dbh=shift;
    if( $GLOBAL_DESTROY ) {
      if( $GLOBAL_DESTROY>1 ) {
	$LOG->(2, "GLOBAL DESTROY $dbh->{$PRIVATE}->{idx}");
      }
      $dbh->SUPER::disconnect;
    } else {
      $LOG->(2, "DESTROY $dbh->{$PRIVATE}->{idx}");
      $dbh->disconnect;
    }
    1;
  }
}

1;
__END__
